nav2_ceres_costaware_smoother/include/nav2_ceres_costaware_smoother/smoother_cost_function.hpp

```diff
@@ -0,0 +1,284 @@
+    const Eigen::Matrix<T, 2, 1> & pt_m,
+    T & r) const
+  {
+    Eigen::Matrix<T, 2, 1> center = arcCenter(
+      pt_m, pt, pt_p,
+      _next_to_last_length_ratio < 0 ? -1 : 1);
+    if (ceres::IsInfinite(center[0])) {
+      return;
+    }
+    T turning_rad = (pt - center).norm();
+    T ki_minus_kmax = (T)1.0 / turning_rad - _params.max_curvature;
+
+    if (ki_minus_kmax <= (T)EPSILON) {
+      return;
+    }
+
+    r += (T)weight * ki_minus_kmax * ki_minus_kmax;  // objective function value
+  }
+
+  /**
+   * @brief Cost function derivative term for steering away changes in pose
+   * @param weight Weight to apply to function
+   * @param xi Point Xi for evaluation
+   * @param xi_original original point Xi for evaluation
+   * @param r Residual (cost) of term
+   */
+  template<typename T>
+  inline void addDistanceResidual(
+    const double & weight,
+    const Eigen::Matrix<T, 2, 1> & xi,
+    const Eigen::Matrix<T, 2, 1> & xi_original,
+    T & r) const
+  {
+    r += (T)weight * (xi - xi_original).squaredNorm();  // objective function value
+  }
+
+  /**
+   * @brief Cost function term for steering away from costs
+   * @param weight Weight to apply to function
+   * @param value Point Xi's cost'
+   * @param params computed values to reduce overhead
+   * @param r Residual (cost) of term
+   */
+  template<typename T>
+  inline void addCostResidual(
+    const double & weight,
+    const Eigen::Matrix<T, 2, 1> & pt,
+    const Eigen::Matrix<T, 2, 1> & pt_p1,
+    const Eigen::Matrix<T, 2, 1> & pt_m1,
+    T & r) const
+  {
+    if (_params.cost_check_points.empty()) {
+      Eigen::Matrix<T, 2,
+        1> interp_pos = (pt - _costmap_origin.template cast<T>()) / (T)_costmap_resolution;
+      T value;
+      _costmap_interpolator->Evaluate(interp_pos[1] - (T)0.5, interp_pos[0] - (T)0.5, &value);
+      r += (T)weight * value * value;  // objective function value
+    } else {
+      Eigen::Matrix<T, 2, 1> dir = tangentDir(
+        pt_m1, pt, pt_p1,
+        _next_to_last_length_ratio < 0 ? -1 : 1);
+      dir.normalize();
+      if (((pt_p1 - pt).dot(dir) < (T)0) != _reversing) {
+        dir = -dir;
+      }
+      Eigen::Matrix<T, 3, 3> transform;
+      transform << dir[0], -dir[1], pt[0],
+        dir[1], dir[0], pt[1],
+      (T)0, (T)0, (T)1;
+      for (size_t i = 0; i < _params.cost_check_points.size(); i += 3) {
+        Eigen::Matrix<T, 3, 1> ccpt((T)_params.cost_check_points[i],
+          (T)_params.cost_check_points[i + 1], (T)1);
+        auto ccpt_world = (transform * ccpt).template block<2, 1>(0, 0);
+        Eigen::Matrix<T, 2,
+          1> interp_pos = (ccpt_world - _costmap_origin.template cast<T>()) /
+          (T)_costmap_resolution;
+        T value;
+        _costmap_interpolator->Evaluate(interp_pos[1] - (T)0.5, interp_pos[0] - (T)0.5, &value);
+
+        r += (T)weight * (T)_params.cost_check_points[i + 2] * value * value;
+      }
+    }
+  }
+
+  const Eigen::Vector2d _original_pos;
```

Comment:
trailing underscore please

Reply:
Actually, you can just make the small number of `variable_`'s into `_variables`'s, it'll be easier for you than to change all of these all over the place for just 3